### PR TITLE
Create docs for Button core, restart, template

### PR DESCRIPTION
**Pull request by jesserockz**
## Description:

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome](https://github.com/esphome/esphome) with YAML changes (if applicable):** esphome/esphome#2824

## Checklist:

  - [x] Branch: `next` is for changes and new documentation that will go public with the next ESPHome release. Fixes, changes and adjustments for the current release should be created against `current`.
  - [x] Link added in `/index.rst` when creating new documents for new components or cookbook.

### CODE REVIEW

File: components/button/index.rst [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+All buttons in ESPHome have a name and an optional icon.
+
+.. code-block:: yaml
+
+    # Example button configuration
+    name: Livingroom Lazy Mood
+    id: my_button
+
+    # Optional variables:
+    icon: "mdi:emoticon-outline"
```
Comment (oxan) [inline components/button/index.rst]: Usually we show the parent here as well:
```yaml
button:
  - platform: ..
    name: Livingroom Lazy Mood
    ...
```

File: components/button/index.rst [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+- **disabled_by_default** (*Optional*, boolean): If true, then this entity should not be added to any client's frontend,
+  (usually Home Assistant) without the user manually enabling it (via the Home Assistant UI).
+- **entity_category** (*Optional*, string): The category of the entity.
+  See https://developers.home-assistant.io/docs/core/entity/#generic-properties
+  for a list of available options. Set to ``""`` to remove the default entity category.
+
+Automations:
+
+- **on_value** (*Optional*, :ref:`Automation <automation>`): An automation to perform
+  when a new value is published. See :ref:`select-on_value`.
```
Comment (oxan) [inline components/button/index.rst]: copy/paste error.

File: components/button/index.rst [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+MQTT Options:
```
Comment (oxan) [inline components/button/index.rst]:
```suggestion
MQTT options:
```

File: components/button/index.rst [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+.. code-block:: yaml
+
+    button:
+      - platform: template
+        # ...
+        on_press:
+          then:
+            - logger.log: Button Pressed
+
+Configuration variables: See :ref:`Automation <automation>`.
```
Comment (oxan) [inline components/button/index.rst]:
```suggestion
Configuration variables: see :ref:`Automation <automation>`.
```

File: components/button/index.rst [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+This is an :ref:`Action <config-action>` for pressing a button in an Automation.
+
+.. code-block:: yaml
+
+    - button.press: my_button
+
+Configuration variables:
+
+- **id** (**Required**, :ref:`config-id`): The ID of the select to set.
```
Comment (oxan) [inline components/button/index.rst]: copy/paste error.

File: components/button/index.rst [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+.. _button-lambda_calls:
+
+lambda calls
+************
+
+From :ref:`lambdas <config-lambda>`, you can press a button.
+
+- ``press()``: Set the select option.
```
Comment (oxan) [inline components/button/index.rst]: copy/paste error.

**Comment by jesserockz** [comment]
Thanks @oxan

File: components/button/index.rst [hunk 7]
```diff
@@ -0,0 +1,118 @@
+Button Component
+================
+
+.. seo::
+    :description: Instructions for setting up button components in ESPHome.
+    :image: folder-open.svg
+
+ESPHome has support for components to create a button entity. A button entity is
+basically a momentary switch with no state and can be triggered by either yaml or
+the user/frontend.
```
Comment (frenck) [inline components/button/index.rst]:
```suggestion
ESPHome has support for components to create a button entity. A button entity is
basically a momentary switch with no state and can be triggered by either YAML or
the user/frontend.
```

**Review by frenck — APPROVED** [review]

**Pull request merged**